File: scripts/train-limiter-policy.mjs
import fs from "fs";
import path from "path";
import { readNdjson, sanitizeStyle } from "../lib/ml-policy-dataset.mjs";
import {
  buildLimiterSamples,
  trainLimiterPolicyModel,
  saveLimiterPolicyModel,
  LIMITER_POLICY_PATH
} from "../lib/limiter-policy.mjs";

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) continue;
    const key = item.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    out[key] = next;
    i += 1;
  }
  return out;
}

function ensureDir(dirPath) {
  if (!fs.existsSync(dirPath)) fs.mkdirSync(dirPath, { recursive: true });
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const style = sanitizeStyle(String(args.style || ""));
  const defaultFeedbackPath = style
    ? path.resolve(process.cwd(), "data", "training", "styles", style, "feedback.ndjson")
    : path.resolve(process.cwd(), "data", "training", "feedback.ndjson");
  const feedbackPath = path.resolve(String(args["feedback-file"] || defaultFeedbackPath));
  const outputModel = path.resolve(String(args["output-model"] || LIMITER_POLICY_PATH));
  const summaryFile = String(args["summary-file"] || "").trim();
  const rows = readNdjson(feedbackPath);
  const samples = buildLimiterSamples(rows, { style });
  const res = trainLimiterPolicyModel({ samples });
  if (res.status !== "trained") {
    console.log(`limiter_training_skipped reason=${res.reason} samples=${res.sampleCount}`);
    return;
  }

  const meta = {
    createdAt: new Date().toISOString(),
    sampleCount: res.sampleCount,
    styleScope: style || "all",
    feedbackFile: path.relative(process.cwd(), feedbackPath).replace(/\\/g, "/"),
    version: "limiter-policy-v1"
  };
  const saved = saveLimiterPolicyModel({ model: res.model, meta, modelPath: outputModel });
  console.log(
    `limiter_policy_trained model=${path.relative(process.cwd(), saved).replace(/\\/g, "/")} samples=${res.sampleCount}`
  );

  if (summaryFile) {
    const target = path.resolve(summaryFile);
    ensureDir(path.dirname(target));
    fs.writeFileSync(
      target,
      JSON.stringify(
        {
          status: "trained",
          modelPath: path.relative(process.cwd(), saved).replace(/\\/g, "/"),
          sampleCount: res.sampleCount,
          styleScope: style || "all"
        },
        null,
        2
      ),
      "utf8"
    );
  }
}

main();

File: scripts_py/train_limiter_policy.py
import json
import os
from datetime import datetime, timezone
from pathlib import Path

from common_pipeline import parse_args
from limiter_policy import build_samples, read_ndjson, save_model, train_linear


def main():
    args = parse_args(os.sys.argv[1:])
    cwd = os.getcwd()
    style = str(args.get("style") or "").strip().lower()
    default_feedback = (
        Path(cwd) / "data" / "training" / "styles" / style / "feedback.ndjson"
        if style
        else Path(cwd) / "data" / "training" / "feedback.ndjson"
    )
    feedback_path = Path(str(args.get("feedback-file") or default_feedback)).resolve()
    output_model = Path(str(args.get("output-model") or "models/limiter-policy-v1-py.json")).resolve()
    summary_file = str(args.get("summary-file") or "").strip()
    rows = read_ndjson(str(feedback_path))
    samples = build_samples(cwd, rows, style=style)
    res = train_linear(samples, ridge=1e-6)
    if res.get("status") != "trained":
        print(f"limiter_training_skipped reason={res.get('reason')} samples={res.get('sampleCount', 0)}")
        return

    payload = {
        "modelType": "linear_py_v1",
        "intercept": float(res["intercept"]),
        "weights": [float(v) for v in res["weights"]],
        "meta": {
            "createdAt": datetime.now(timezone.utc).isoformat(),
            "sampleCount": int(res["sampleCount"]),
            "styleScope": style or "all",
            "feedbackFile": str(feedback_path.relative_to(Path(cwd))).replace("\\", "/"),
            "version": "limiter-policy-v1-py",
            "trainer": "python",
        },
    }
    saved = save_model(str(output_model), payload)
    print(
        f"limiter_policy_trained model={str(saved.relative_to(Path(cwd))).replace('\\', '/')} samples={res['sampleCount']}"
    )
    if summary_file:
        sp = Path(summary_file).resolve()
        sp.parent.mkdir(parents=True, exist_ok=True)
        with open(sp, "w", encoding="utf-8") as f:
            json.dump(
                {
                    "status": "trained",
                    "modelPath": str(saved.relative_to(Path(cwd))).replace("\\", "/"),
                    "sampleCount": int(res["sampleCount"]),
                    "styleScope": style or "all",
                },
                f,
                indent=2,
                ensure_ascii=False,
            )


if __name__ == "__main__":
    main()
